src/reducer.ts

```diff
@@ -132,6 +133,7 @@ export default function reducer(
         ...state,
         model: updateCellValue(state.model, active, cellData),
         lastChanged: active,
+        lastUpdateDate: new Date(),
```

Comment:
Why is this needed?

Reply:
Hi Iddan,

First off, it's commonly advised against including objects directly in the dependency array of a `useEffect`. This is because objects are compared by reference, not by value. What this means is that even if the content of the object hasn't changed, if a new object is created on each render, the effect will still run on every render because the references are different.

So, although comparing object references directly might seem like a good idea, it's actually not recommended.

In the case of react-spreadsheet, the code below can lead to an infinite loop due to shallow comparisons made within the useEffect body and the dependency array.

```tsx
<Spreadsheet
  data={data}
  onChange={(newData: any) => {
    console.log(newData);
    setData(newData);
  }}
/>
```

```tsx
React.useEffect(() => {
  if (state.model.data !== prevDataRef.current) {
    // Call onChange only if the data changes internally
    if (state.model.data !== props.data) {
      onChange(state.model.data);
    }
  }
  prevDataRef.current = state.model.data;
}, [state.model.data, onChange, props.data]);
```

![2024-05-1012-03-19-ezgif com-video-to-gif-converter](https://github.com/iddan/react-spreadsheet/assets/48513387/92076876-0a33-46a1-a0ee-a7ff67c500ef)

To illustrate this point further, you can try changing the second block to:

```tsx
React.useEffect(() => {
  console.log("Use Effect called");
}, [state.model.data]);
```
You'll notice that even though the object might appear to be the same instance, its reference changes after every rendering. This causes the console log to be called every time there is a modification.

**To summarize**:

1. `state.model.data` changes its reference every time `setCellData` is called.
2. Since the reference changes, the `useEffect` body is executed.
3. The first `if` condition is verified as the reference is always different from the one previously stored.
4. The second `if` condition is verified because there is no guarantee `props.data` has a reference equal to the new `state.model.data`.
5. The `onChange` function calls the `setData` function within the `SpreadSheet` object in `App.tsx`.
6. The loop restarts without an end.

**Solution**

Adding a new `lastUpdateDate` property addresses this issue while avoiding resource-intensive deep nested comparisons.

With this addition, we track the last time `setCellData` is executed. If the data displayed in the UI has a different `lastUpdateDate` value, the `useEffect` is called, and the displayed information is updated accordingly.

Reply:
`React.useEffect(() => {

}, [JSON.stringify(state.modal.data)]);
`

Why not use just a string compare? For the huge datasets, it might be too slow though, but there is other libraries that can handle stringifying of an object more performant. Comparing a string, instead of having to traverse a huge object-graph is way easier. 

The lastUpdate date check can probably cause some bad sideeffects as well.

Reply:
Thank you @EdoardoGruppi for the detailed writing. I am not aware object reference compare is not recommended by any official React resources. I think the real problem is the array reference changes which I don't see why it should happen more than once per mutation to the data.

Reply:
@EdoardoGruppi are you using the latest version of the code? The controlled storybook story works correctly and it uses the same code as you posted.

Reply:
@thsorens in the same spirit of React, I try to avoid unsimple comparisons on data.

Reply:
Hi @iddan, 
thanks for your responses. 

I confirm that I am using the latter version of the code. To be 100% sure, I re-did the test by cloning the repo from scratch, but nothing changed. You can find a sandbox at this [link](https://codesandbox.io/p/devbox/practical-nova-m3fvr9?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clx8rkyzf00083j6g1t9vo7xo%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clx8rkyzf00023j6gzo73nhjs%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clx8rkyzf00063j6go51azxm7%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clx8rkyzf00073j6g1q6je5f3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clx8rkyzf00023j6gzo73nhjs%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx8rkyze00013j6gtwfsg08b%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252F.codesandbox%252Ftemplate.json%2522%257D%255D%252C%2522id%2522%253A%2522clx8rkyzf00023j6gzo73nhjs%2522%252C%2522activeTabId%2522%253A%2522clx8rkyze00013j6gtwfsg08b%2522%257D%252C%2522clx8rkyzf00073j6g1q6je5f3%2522%253A%257B%2522id%2522%253A%2522clx8rkyzf00073j6g1q6je5f3%2522%252C%2522activeTabId%2522%253A%2522clx8rl3tj002c3j6gcnx2qud3%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clx8rl3tj002c3j6gcnx2qud3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clx8rkyzf00063j6go51azxm7%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx8rkyzf00033j6g87naf9rf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clx8rkyzf00043j6g0bgbih2z%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clx8ql7wq001rdaipc3imepq5%2522%257D%252C%257B%2522id%2522%253A%2522clx8rkyzf00053j6g3vmx32em%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clx8qmcmk001tdaip2f4k69en%2522%257D%255D%252C%2522id%2522%253A%2522clx8rkyzf00063j6go51azxm7%2522%252C%2522activeTabId%2522%253A%2522clx8rkyzf00033j6g87naf9rf%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) to check the results I am getting. To see the logs, you just need to open the console in the embedded CodeSandbox browser.

You're correct that React doesn't explicitly advise against using non-primitive values inside the dependency array. However, many developers recommend avoiding this practice because it can lead to mistakes and unintended re-renders.

I am confident that this is the exact problem with the react-spreadsheet library. In one of my initial commits, I used a function to deep compare objects, which solved the issue. The same error can be highlighted when adding a log inside the outer if statement. The comparison `state.model.data !== prevDataRef.current` returns true endlessly every time a cell changes its value.

```tsx
React.useEffect(() => {
  if (state.model.data !== prevDataRef.current) {
      console.log('state.model.data is not equal to prevDataRef.current')
  }
  prevDataRef.current = state.model.data;
}, [state.model.data]);
```

It's worth mentioning that I later replaced that function to avoid slowdowns with a comparison made on a new lastUpdatedDate property of the state object. I think that using the date is a nice way to solve the issue since: it doesn't add much code or memory occupation, makes the comparison very fast, and provides additional information that may be useful for future features.

Reply:
@iddan : I also found a different weird behavior in the storybook version. Add some random text in a cell. Put the focus on another cell. Then go back to the cell, type 1 (the number 1). The focus will automatically disappear, so it will not be possible to continue writing in the cell. It might be related to the grid running in storybook, since i cant reproduce it in the "demo" version. (Running chrome)